File: connectors/javascript/tests/integration/core/C07_Heartbeat.test.ts
/**
 * C-07: Heartbeat Automatic Handling Tests
 *
 * Tests that the Connector automatically sends heartbeats
 * and maintains connections over time.
 */

import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import { BaseIntegrationTest } from '../helpers/BaseIntegrationTest.js';
import { Connector } from '../../../src/connector.js';
import { Packet } from '../../../src/packet.js';
import { serializeAuthenticateRequest } from '../helpers/TestMessages.js';

describe('C-07: Heartbeat Automatic Handling', () => {
    let testContext: BaseIntegrationTest;

    beforeEach(async () => {
        testContext = new BaseIntegrationTest();
        await testContext['beforeEach']();
    });

    afterEach(async () => {
        await testContext['afterEach']();
    });

    test('C-07-01: Connection maintained over time without disconnection', { timeout: 60000 }, async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('heartbeatUser');

        expect(testContext['connector']!.isConnected).toBe(true);

        // When: Wait 5 seconds without any action (Heartbeat should be sent automatically)
        const startTime = Date.now();
        while (Date.now() - startTime < 5000) {
            testContext['connector']!.mainThreadAction();
            await testContext['delay'](50);
        }

        // Then: Connection should remain active
        expect(testContext['connector']!.isConnected).toBe(true);
        expect(testContext['connector']!.isAuthenticated).toBe(true);
    });

    test('C-07-02: Echo requests work correctly during heartbeat period', async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('echoHeartbeatUser');

        // When: Wait 2 seconds then send echo request
        await testContext['delay'](2000);
        const echoReply = await testContext['echo']('After Heartbeat', 1);

        // Then: Echo should work normally
        expect(echoReply).toBeDefined();
        expect(echoReply.content).toBe('After Heartbeat');

        // Connection should remain active
        expect(testContext['connector']!.isConnected).toBe(true);
    });

    test('C-07-03: Short heartbeat interval works correctly', { timeout: 10000 }, async () => {
        // Given: Short heartbeat interval (1 second)
        const connector = new Connector();
        connector.init({
            requestTimeoutMs: 5000,
            heartbeatIntervalMs: 1000 // 1 second heartbeat
        });
        testContext['connector'] = connector;

        // When: Connect and authenticate
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('shortHeartbeatUser');

        // Wait 3 seconds (3 heartbeats will be sent)
        const startTime = Date.now();
        while (Date.now() - startTime < 3000) {
            connector.mainThreadAction();
            await testContext['delay'](50);
        }

        // Then: Connection should be maintained
        expect(connector.isConnected).toBe(true);

        // Echo should work normally
        const echoReply = await testContext['echo']('Short Interval Test', 1);
        expect(echoReply.content).toBe('Short Interval Test');
    });

    test('C-07-04: Message transmission works correctly during heartbeat', { timeout: 15000 }, async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('transmitUser');

        // When: Send echo requests periodically for 10 seconds (along with heartbeat)
        for (let i = 1; i <= 5; i++) {
            const echoReply = await testContext['echo'](`Message ${i}`, i);
            expect(echoReply.content).toBe(`Message ${i}`);

            await testContext['delay'](2000); // 2 seconds between messages
        }

        // Then: All messages should be sent successfully and connection maintained
        expect(testContext['connector']!.isConnected).toBe(true);
    });

    test('C-07-05: OnDisconnect event does not trigger during normal operation', { timeout: 15000 }, async () => {
        // Given: Connected and authenticated
        await testContext['createStageAndConnect']();
        await testContext['authenticate']('noDisconnectUser');

        let disconnectTriggered = false;
        testContext['connector']!.onDisconnect = () => {
            disconnectTriggered = true;
        };

        // When: Wait 5 seconds
        const startTime = Date.now();
        while (Date.now() - startTime < 5000) {
            testContext['connector']!.mainThreadAction();
            await testContext['delay'](50);
        }

        // Then: OnDisconnect should not be triggered
        expect(disconnectTriggered).toBe(false);
        expect(testContext['connector']!.isConnected).toBe(true);
    });

    test('C-07-06: Multiple connectors can maintain heartbeat simultaneously', { timeout: 30000 }, async () => {
        // Given: Create 3 connectors
        const stage1 = await testContext['testServer'].createStage('TestStage');
        const stage2 = await testContext['testServer'].createStage('TestStage');
        const stage3 = await testContext['testServer'].createStage('TestStage');

        const connector1 = new Connector();
        const connector2 = new Connector();
        const connector3 = new Connector();

        try {
            connector1.init({ requestTimeoutMs: 5000, heartbeatIntervalMs: 10000 });
            connector2.init({ requestTimeoutMs: 5000, heartbeatIntervalMs: 10000 });
            connector3.init({ requestTimeoutMs: 5000, heartbeatIntervalMs: 10000 });

            const wsUrl = testContext['testServer'].wsUrl;
            await connector1.connect(wsUrl, stage1.stageId, stage1.stageType);
            await connector2.connect(wsUrl, stage2.stageId, stage2.stageType);
            await connector3.connect(wsUrl, stage3.stageId, stage3.stageType);

            // Authenticate all connectors using protobuf format
            const authPayload1 = serializeAuthenticateRequest({ userId: 'user1', token: 'token1' });
            const authPayload2 = serializeAuthenticateRequest({ userId: 'user2', token: 'token2' });
            const authPayload3 = serializeAuthenticateRequest({ userId: 'user3', token: 'token3' });

            await connector1.authenticate(Packet.fromBytes('AuthenticateRequest', authPayload1));
            await connector2.authenticate(Packet.fromBytes('AuthenticateRequest', authPayload2));
            await connector3.authenticate(Packet.fromBytes('AuthenticateRequest', authPayload3));

            expect(connector1.isAuthenticated).toBe(true);
            expect(connector2.isAuthenticated).toBe(true);
            expect(connector3.isAuthenticated).toBe(true);

            // When: Wait 5 seconds (all connectors' heartbeats should work)
            const startTime = Date.now();
            while (Date.now() - startTime < 5000) {
                connector1.mainThreadAction();
                connector2.mainThreadAction();
                connector3.mainThreadAction();
                await testContext['delay'](50);
            }

            // Then: All connectors should remain connected
            expect(connector1.isConnected).toBe(true);
            expect(connector2.isConnected).toBe(true);
            expect(connector3.isConnected).toBe(true);
        } finally {
            connector1.disconnect();
            connector2.disconnect();
            connector3.disconnect();
        }
    });
});

File: connectors/cpp/tests/integration/core/c07_heartbeat_test.cpp
#include "../base_integration_test.hpp"
#include <gtest/gtest.h>

using namespace playhouse;
using namespace playhouse::test;

/// C-07: Heartbeat Tests
/// Verifies automatic heartbeat mechanism
class C07_HeartbeatTest : public BaseIntegrationTest {};

TEST_F(C07_HeartbeatTest, Heartbeat_AutomaticallySent_KeepsConnectionAlive) {
    // Given: Connected with short heartbeat interval
    config_.heartbeat_interval_ms = 1000;  // 1 second for testing
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    ASSERT_TRUE(CreateStageAndConnect());

    // When: Wait for multiple heartbeat intervals
    std::this_thread::sleep_for(std::chrono::milliseconds(3000));

    // Keep calling MainThreadAction to process heartbeats
    for (int i = 0; i < 30; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // Then: Connection should still be alive
    EXPECT_TRUE(connector_->IsConnected()) << "Connection should remain alive with heartbeats";
}

TEST_F(C07_HeartbeatTest, Heartbeat_LongInterval_ConnectionStable) {
    // Given: Connected with long heartbeat interval
    config_.heartbeat_interval_ms = 10000;  // 10 seconds
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    ASSERT_TRUE(CreateStageAndConnect());

    // When: Wait and process callbacks
    for (int i = 0; i < 50; i++) {
        connector_->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // Then: Connection should be stable
    EXPECT_TRUE(connector_->IsConnected());
}

TEST_F(C07_HeartbeatTest, Heartbeat_DoesNotInterfereWithNormalMessages) {
    // Given: Connected with heartbeat enabled
    config_.heartbeat_interval_ms = 2000;
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    ASSERT_TRUE(CreateStageAndConnect());

    // When: Send normal messages during heartbeat period
    std::string echo_data = "{\"content\":\"Test during heartbeat\",\"sequence\":1}";
    Bytes payload(echo_data.begin(), echo_data.end());
    auto packet = Packet::FromBytes("EchoRequest", std::move(payload));

    auto future = connector_->RequestAsync(std::move(packet));

    // Then: Normal messages should work fine
    try {
        auto response = WaitWithMainThreadAction(future, 5000);
        EXPECT_EQ(response.GetErrorCode(), 0) << "Normal message should succeed during heartbeat";
    } catch (const std::exception& e) {
        FAIL() << "Message failed during heartbeat: " << e.what();
    }
}

TEST_F(C07_HeartbeatTest, Heartbeat_ConfiguredInterval_RespectedByConnector) {
    // Given: Specific heartbeat interval configured
    config_.heartbeat_interval_ms = 5000;  // 5 seconds
    connector_ = std::make_unique<Connector>();
    connector_->Init(config_);

    // When: Connect to server
    ASSERT_TRUE(CreateStageAndConnect());

    // Then: Connection should establish successfully with configured heartbeat
    EXPECT_TRUE(connector_->IsConnected());
    // The actual heartbeat timing is internal, but connection should work
}
